src/main/java/io/stargate/sgv2/jsonapi/api/model/command/clause/filter/LogicalExpression.java

```diff
@@ -25,10 +26,67 @@ public String getOperator() {
     }
   }
 
-  private final LogicalOperator logicalRelation;
+  private LogicalOperator logicalRelation;
   private int totalComparisonExpressionCount;
   private int totalIdComparisonExpressionCount;
 
+  /** This method will flip the operators and operand if logical operator is not */
+  public void traverseForNot(LogicalExpression parent) {
+    List<LogicalExpression> tempLogicalExpressions = new ArrayList<>(logicalExpressions);
+    for (LogicalExpression logicalExpression : tempLogicalExpressions) {
+      logicalExpression.traverseForNot(this);
+    }
+
+    Iterator<LogicalExpression> iterator = logicalExpressions.iterator();
+    while (iterator.hasNext()) {
+      LogicalExpression logicalExpression = iterator.next();
+      if (logicalExpression.logicalRelation == LogicalOperator.NOT) {
+        iterator.remove();
+        this.totalComparisonExpressionCount =
+            this.totalComparisonExpressionCount - logicalExpression.totalComparisonExpressionCount;
+        this.totalIdComparisonExpressionCount =
+            this.totalIdComparisonExpressionCount
+                - logicalExpression.totalIdComparisonExpressionCount;
+      }
+    }
+
+    if (logicalRelation == LogicalOperator.NOT) {
+      flip();
+      addToParent(parent);
+    }
+  }
+
+  private void addToParent(LogicalExpression parent) {
+    logicalExpressions.stream().forEach(expression -> parent.addLogicalExpression(expression));
+    if (comparisonExpressions.size() > 1) {
+      // Multiple conditions in not, after push down will become or
+      final LogicalExpression orLogic = LogicalExpression.or();
+      comparisonExpressions.stream()
+          .forEach(comparisonExpression -> orLogic.addComparisonExpression(comparisonExpression));
+      parent.addLogicalExpression(orLogic);
+    } else {
+      if (comparisonExpressions.size() == 1) {
+        // Unary not, after push down will become additional condition
+        parent.addComparisonExpression(comparisonExpressions.get(0));
+      }
+    }
+    comparisonExpressions.clear();
+  }
+
+  private void flip() {
+    if (logicalRelation == LogicalOperator.AND) {
+      logicalRelation = LogicalOperator.OR;
+    } else if (logicalRelation == LogicalOperator.OR) {
+      logicalRelation = LogicalOperator.AND;
+    }
```

Comment:
Is it possible this could be called for `NOT`? And if so, is "do nothing" what should be done (if it shouldn't be called, should probably throw `IllegalStateException` in `else` case.

Reply:
We resolve depth first handled `NOT` by flipping the operators and moved the expression to the parent logical expression. Because of that flip() will not have any `NOT` logical expression below it.

Reply:
Ok, thanks.